src/Object/ObjectFactory.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+/*
+ * This file is part of the zenstruck/foundry package.
+ *
+ * (c) Kevin Bond <kevinbond@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Zenstruck\Foundry\Object;
+
+use Zenstruck\Foundry\BaseFactory;
+use Zenstruck\Foundry\FactoryCollection;
+use Zenstruck\Foundry\Instantiator;
+use Zenstruck\Foundry\LazyValue;
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ *
+ * @template T of object
+ * @extends BaseFactory<T>
+ *
+ * @phpstan-type CallableInstantiator Instantiator<T>|\Closure(Parameters,class-string<T>):T
+ */
+abstract class ObjectFactory extends BaseFactory
+{
+    /** @var CallableInstantiator */
+    private \Closure|Instantiator $instantiator;
+
+    /** @var array<callable(Parameters,class-string<T>):Parameters> */
+    private array $beforeInstantiate = [];
+
+    /** @var array<callable(T,Parameters):void> */
+    private array $afterInstantiate = [];
+
+    /**
+     * @return class-string<T>
+     */
+    public static function class(): string
+    {
+        return static::getClass();
+    }
+
+    /**
+     * @return class-string<T>
+     */
+    abstract protected static function getClass(): string;
```

Comment:
not sure what to do here: this is made here to prevent BC breaks (a public method is need to list factories in `\Zenstruck\Foundry\Bundle\Maker\Factory\FactoryClassMap`)

Reply:
Why can't we use `::class()` above in the `FactoryClassMap`?

Reply:
:thinking: hmmm OK what  I did was wrong! I've overridden `self::class()` in `ModelFactory` now. When users will move from `ModelFactory` to one of `PersistentObjectFactory` or `ObjectFactory`, they'll also need to mode from `protected function getClass()` to `public function class()`